src/api/lock.jl

```diff
@@ -0,0 +1,14 @@
+const liblock = ReentrantLock()
+
+# Try to acquire the lock (test-test-set) and close if successful
+# Otherwise, defer finalization
+# https://github.com/JuliaIO/HDF5.jl/issues/1048
+function try_close_finalizer(x)
+    if !islocked(liblock) && trylock(liblock) do
+        close(x)
+        true
```

Comment:
I don't think we need this here, as it is handled by `trylock`
```suggestion
```

Reply:
I think we do need this. The two argument `trylock` above does `return f()`. The result of `true && nothing` is `nothing`, and that will be provided to the `if` statement. If the `if` statement receives a `nothing`, an error will be thrown.

```julia
julia> if true && nothing
       else
           println("Hello")
       end
ERROR: TypeError: non-boolean (Nothing) used in boolean context
Stacktrace:
 [1] top-level scope
   @ REPL[27]:1
```

Reply:
Oh, I didn't realize it returns `f()`.